File: 931. Minimum Falling Path Sum.py
# Submitted by : Aryan Singh_RN12MAY2023
# Time Complexity : O(n^2)
# Space Complexity : Average : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    '''
    - Repetetive subproblems were detected during exhaustive approach, so dp being used
    - during iteration to a specific value, all connected values in next row(c, c+1, c-1)
    - are compared and minimum sum is stored in memory.
    '''
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        n = len(matrix)
        if n > 1:
            for r in range(n -1)[::-1]:
                for c in range(n):
                    if c == 0:
                        matrix[r][c] = matrix[r][c] + min(matrix[r+1][c], matrix[r+1][c+1])
                    elif c == n - 1:
                        matrix[r][c] = matrix[r][c] + min(matrix[r+1][c], matrix[r+1][c-1])
                    else:
                        matrix[r][c] = matrix[r][c] + min(matrix[r+1][c], matrix[r+1][c+1], matrix[r+1][c-1])
        minimum = matrix[0][0]
        for c in range(n):
            minimum = min(minimum, matrix[0][c])
        return minimum

File: _931_MinimumFallingPathSum.java
// space complexity - o(n*2)
    	//time - o(n*2)
      //Ran on leeetcode successfully : Yes
      // Problem faced  : No
    	//Approach : bottom up iterative solution. idea is to calculate the current cell value to be equal to the
       // min of above cell and diagonal left and right cell + current cell value
       

class Solution {
    public int minFallingPathSum(int[][] matrix) {

      int[][] dp = new int[matrix.length + 1][matrix[0].length + 2];

      for(int i=0; i<dp.length; i++){
        for(int j = 0 ; j< dp[0].length; j++){

          if(i == 0 ){
            dp[i][j] = 0;
          }
          else if( j == 0 || j == dp[0].length-1){
            dp[i][j] = Integer.MAX_VALUE;
          }
          else{
            dp[i][j] = matrix[i-1][j-1] + Math.min(dp[i-1][j-1], Math.min(dp[i-1][j], dp[i-1][j+1]));
          }
        }
      }

      int res = Integer.MAX_VALUE;
      for(int j=1; j<dp[0].length-1; j++){
        res = Math.min(res, dp[matrix.length][j]);
      }
      return res;
    }
}
